### PR TITLE
Adding analog input read functionality 

### PR DESCRIPTION
**Pull request by jerlfan**
Adding a 0D viewer to monitor the analog inputs of an Arduino UNO card. 
This version is working properly. Need to add functionality for changing the units and include the possibility of separated viewers for each channel.

### CODE REVIEW

File: src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+from pymodaq.utils.data import DataFromPlugins, DataToExport
+from pymodaq.control_modules.viewer_utility_classes import DAQ_Viewer_base, comon_parameters, main
+from pymodaq.utils.parameter import Parameter
+
+from typing import Optional
+
+from pymodaq_plugins_arduino.hardware.arduino_telemetrix import Arduino
+from pymodaq_plugins_arduino.utils import Config
+
+config=Config()
```
Comment (seb5g) [inline src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py]: add two lines in between config and the class declaration

Comment (seb5g) [reply src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py]: add space before and after the equal sign

File: src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        """
+
+        self.ini_detector_init(slave_controller=controller)
+
+        if self.is_master:
+            self.controller = Arduino(com_port=self.settings['com_port'])  # instantiate you driver with whatever arguments are needed
+
+
+        # TODO for your custom plugin (optional) initialize viewers panel with the future type of data
+        self.dte_signal_temp.emit(DataToExport(name='AnalogInput',
```
Comment (seb5g) [inline src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py]: you don't really have to emit the temp signal here if the type of data emission is the same as the declared plugin (here 0D for 0D data so the viewer will already be correctly set)

File: src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            if 'AI' in param.name():
+                if param['ch']:
+                    self.controller.set_analog_input(int(param.name()[2:3]))
+                    data_tot.append(np.array([self.controller.analog_pin_values_input[int(param.name()[2:3])]]))
+                else:
+                    data_tot.append(np.array([0]))
+
+        self.dte_signal.emit(DataToExport(name='Analog Input',
+                                          data=[DataFromPlugins(name='AI', data=data_tot,
+                                                                dim='Data0D', labels=['AI0', 'AI1', 'AI2', 'AI3', 'AI4',
```
Comment (seb5g) [inline src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py]: if I'm correct, you won't have alsways the 5 AI channels activated therefore the labels won't always be as you wrote. You could construct this labels list in the same way you construct data_tot, ie appending the label string (param name or other) if param['ch'] is true

File: src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                                                                             'AI5'])]))
+
+
+    def callback(self):
```
Comment (seb5g) [inline src/pymodaq_plugins_arduino/daq_viewer_plugins/plugins_0D/daq_0Dviewer_Analog.py]: if unused, remove the method

File: src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py [hunk 5]
```diff
@@ -33,13 +34,44 @@ def set_pins_output_to(self, value: int):
...
+        Used as a callback function to read the value of the analog inputs.
+        Data[0]: pin_type (not used here)
+        Data[1]: pin_number: i.e. 0 is A0 etc.
+        Data[2]: pin_value: an integer between 0 and 1023 (with an Arduino UNO)
+        Data[3]: raw_time_stamp (not used here)
+        :param data: a list in which are loaded the acquisition parameter analog input
+        :return: a dictionary with the following structure {pin_number(int):pin_value(int)}
+        With an arduino up to 6 analog input might be interrogated at the same time
+        """
+        self.analog_pin_values_input[data[1]] = data[2]*5.0/1023 #Data in Volts
```
Comment (seb5g) [inline src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py]: see here you create it on the fly (wr to my previous remark)

File: src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py [hunk 6]
```diff
@@ -33,13 +34,44 @@ def set_pins_output_to(self, value: int):
...
 
+
     def analog_write_and_memorize(self, pin, value):
         lock.acquire()
         value = self.round_value(value)
         self.analog_write(pin, value)
         self.pin_values_output[pin] = value
         lock.release()
 
+    def read_analog_pin(self,data):
```
Comment (seb5g) [inline src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py]: add a space after the comma

File: src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py [hunk 7]
```diff
@@ -33,13 +34,44 @@ def set_pins_output_to(self, value: int):
...
+        Data[1]: pin_number: i.e. 0 is A0 etc.
+        Data[2]: pin_value: an integer between 0 and 1023 (with an Arduino UNO)
+        Data[3]: raw_time_stamp (not used here)
+        :param data: a list in which are loaded the acquisition parameter analog input
+        :return: a dictionary with the following structure {pin_number(int):pin_value(int)}
+        With an arduino up to 6 analog input might be interrogated at the same time
+        """
+        self.analog_pin_values_input[data[1]] = data[2]*5.0/1023 #Data in Volts
+
+    def set_analog_input(self,pin):
```
Comment (seb5g) [inline src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py]: add space after comma

File: src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py (hunk 5 above)
Comment (seb5g) [reply src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py]: the 5 volt encoded on 1024 bits could depend of hte used arduino. Maybe we should include these settings in the config and ask the user to specify the board they use

**Review by seb5g — CHANGES_REQUESTED** [review]
some changes to be applied! 
However leave it like that for your presentation (you don't want to mess things at the last moment :-))

File: src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py (hunk 5 above)
Comment (jerlfan) [reply src/pymodaq_plugins_arduino/hardware/arduino_telemetrix.py]: I remove this part for now. The idea is to implement a "calibration" methods i.e. convertin the integers in the expected units: example *5/1023 for volts if ten bytes encoding

**Review by seb5g — APPROVED** [review]

**Pull request merged**